src/dag-link.js

```diff
@@ -5,25 +5,52 @@ const mh = require('multihashes')
 // Link represents an IPFS Merkle DAG Link between Nodes.
 class DAGLink {
   constructor (name, size, multihash) {
-    this.name = name
-    this.size = size
+    this._name = name
+    this._size = size
+    this._multihash = null
 
     if (typeof multihash === 'string') {
-      this.multihash = mh.fromB58String(multihash)
+      this._multihash = mh.fromB58String(multihash)
     } else if (Buffer.isBuffer(multihash)) {
-      this.multihash = multihash
+      this._multihash = multihash
     }
+  }
+
+  toString () {
+    const mhStr = mh.toB58String(this.multihash)
+    return `DAGLink <${mhStr} - name: "${this.name}", size: ${this.size}>`
+  }
 
-    this.json = {
+  toJSON () {
+    return {
       name: this.name,
       size: this.size,
-      hash: this.multihash ? mh.toB58String(this.multihash) : undefined
+      hash: this.multihash ? mh.toB58String(this.multihash) : null
     }
   }
 
-  toString () {
-    const mhStr = mh.toB58String(this.multihash)
-    return `DAGLink <${mhStr} - name: "${this.name}", size: ${this.size}>`
+  get name () {
+    return this._name
+  }
+
+  set name (name) {
+    throw new Error(`Can't set property: 'name' is immutable`)
+  }
+
+  get size () {
+    return this._size
+  }
+
+  set size (size) {
+    throw new Error(`Can't set property: 'size' is immutable`)
+  }
+
+  get multihash () {
+    return this._multihash
+  }
+
+  set multihash (multihash) {
+    throw new Error(`Can't set property: 'multihash' is immutable`)
```

Comment:
Wrapping them to ImmutableError doesn't make sense anymore. What we would need to do is:
```javascript
class ImmutableError extends Error {
  constructor(error) {
    super(error)
  }
}
```

So let's just throw an Error.